statsfs/otelstats/subsys_metric_parser.go

```diff
@@ -0,0 +1,199 @@
+//				device paths = /eth0/sub0
+//				metricFileName = latency
+//	Output:
+// 		label = /eth0/sub0
+func (m SubsysMetrics) getMetricLabel(path string) string {
+	metricStr := strings.Split(path, m.SubSystemPath)[1]
+	labelSeg := strings.Split(metricStr, "/")
+	label := strings.Join(labelSeg[:len(labelSeg)-1], "/")
+	return label
+}
+
+// Given a path to a statsfs file, return a MetricInfo struct with label
+// computed by getMetricLabel method and Path being the input path
+// (the path from where the metric could be retrieved)
+// Example:
+//	Input:
+//		m.SubSystemPath = /sys/kernel/stats/net
+//		path = /sys/kernel/stats/net/eth0/sub0/latency
+//	Output:
+//		MetricInfo{
+//			Label: /eth0/sub0
+//			Path: /sys/kernel/stats/net/eth0/sub0/latency
+//		}
+func (m SubsysMetrics) getMetricInfo(path string) MetricInfo {
+	label := m.getMetricLabel(path)
+	metricInfo := MetricInfo{Label: label, Path: path}
+	return metricInfo
+}
+
+// Given a path to a statsfs file, update the metricMap of the corresponding
+// SubsysMetric stuct
+// Example:
+//	Input:
+//		m.SubSystemPath = /sys/kernel/stats/net
+// 		m.SubSystemName = net
+//		m.StatsfsPath = /sys/kernel/stats
+//		path = /sys/kernel/stats/net/eth0/latency
+//	Output:
+//		metricName = net/latency
+//		label = /eth0
+//		new entry in m.Metrics[net/latency]:
+//			MetricInfo{Label: /eth0, Path: /sys/kernel/stats/net/eth0/latency}
+func (m SubsysMetrics) updateMetricMapOneEntry(path string) {
+	metricName := m.getMetricName(path)
+	metricInfo := m.getMetricInfo(path)
+	m.Metrics[metricName] = append(m.Metrics[metricName], metricInfo)
+}
+
+func (m SubsysMetrics) updateMetricMap(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		log.Printf("Failed to walk to file %v\n", path)
+	}
+
+	if !info.IsDir() {
+		m.updateMetricMapOneEntry(path)
+	}
+	return err
+}
+
+func (m SubsysMetrics) constructMetricMap() {
+	err := filepath.Walk(m.SubSystemPath, m.updateMetricMap)
+	if err != nil {
+		log.Printf("Failed to parse metrics for subsystem %v at %v", m.SubSystemName, m.SubSystemPath)
+	}
+}
+
+func (m SubsysMetrics) print() {
+	fmt.Printf("StatsfsPath: %v\n", m.StatsfsPath)
+	fmt.Printf("SubSystemName: %v\n", m.SubSystemName)
+	fmt.Printf("SubSystemPath: %v\n", m.SubSystemPath)
+	fmt.Println("Metrics:")
+	for metricName, labels := range m.Metrics {
+		fmt.Printf("\tmetricName: %v,\n\tinfo: \n", metricName)
+		for _, label := range labels {
+			fmt.Printf("\t\tLabel: %v, Path: %v\n", label.Label, label.Path)
+		}
+	}
+}
+
+// Print prints StatsfsMetrics struct
+func (m StatsfsMetrics) Print() {
+	fmt.Print("\n####################################\n")
+	fmt.Printf("StatsfsPath: %v\n\n", m.StatsfsPath)
+	for subsysName, subsysMetrics := range m.Metrics {
+		fmt.Println("------------------")
+		fmt.Printf("Statsfs metrics for subsystem %v:\n\n", subsysName)
+		subsysMetrics.print()
+	}
+	fmt.Printf("####################################\n\n")
+}
```

Comment:
It's much easier for others to read when this function follows the

`type StatsfsMetrics struct {...}`

definition.